python/amici/ode_export.py

```diff
@@ -843,9 +847,65 @@ def import_from_sbml_importer(self,
              for idx in range(len(si.flux_vector))]
         )
         self._eqs['dxdotdx'] = sp.zeros(si.stoichiometric_matrix.shape[0])
+
+        def dx_dt(x_index, x_Sw):
+            '''
+            Produces the appropriate expression for the first derivative of a
+            species with respect to time, for species that reside in
+            compartments with a constant volume, or a volume that is defined by
+            an assignment or rate rule.
+
+            :param x_index:
+                The index (not identifier) of the species in the variables
+                (generated in "sbml_import.py") that describe the model.
+
+            :param x_Sw:
+                The element-wise product of the row in the stoichiometric
+                matrix that corresponds to the species (row x_index) and the
+                flux (kinetic laws) vector.
+            '''
+            x_id = symbols['species']['identifier'][x_index]
+            if x_id in si.compartment_rules:
+                return x_Sw
+            elif x_id in si.compartment_symbols:
+                raise KeyError('This should not happen. There exists an ODE for an compartment that does not have an associated rate rule.')
+            v_name = si.species_compartment[x_index]
+            if v_name in si.compartment_rules:
+                if (si.compartment_rules[v_name]['type_code']
+                        == libsbml.SBML_ASSIGNMENT_RULE):
+                    v = si.compartment_rules[v_name]['formula']
+                    dv_dt = v.diff(si.amici_time_symbol)
+                    dv_dx = v.diff(x_id)
+                    # not sure if parentheses here are necessary...
+                    return ((x_Sw) - (dv_dt)*(x_id))/((dv_dx)*(x_id) + (v))
+                elif (si.compartment_rules[v_name]['type_code']
+                          == libsbml.SBML_RATE_RULE):
+                    t = si.amici_time_symbol
```

Comment:
this looks complicated, would be nice to have a short reminder (as comment in the code) why we need to do this and, if necessary, how we are doing it.

Reply:
I've added a description in the code. I can add the derivation of the equations to the AMICI documentation, is there a suitable page already?